### ISSUE TITLE
Running F# scripts in unix without proper file extensions results in strange behaviour

### ISSUE THREAD

**Issue by eiriktsarpalis** [dotnet/fsharp#1241]
In any unix environment with F# 4.0 installed, running

``` bash
echo '#!/usr/bin/env fsharpi\nprintfn "Hello, World!"' > hello
chmod +x hello
./hello
```

Results in the following behaviour:

```
[12:43] eirik@eudoxos: ~/devel> ./hello

F# Interactive for F# 4.0 (Open Source Edition)
Freely distributed under the Apache 2.0 Open Source License

For help type #help;;

[Loading /home/eirik/devel/hello]


error FS0226: The file extension of '/home/eirik/devel/hello' is not recognized. Source files must have extension .fs, .fsi, .fsx, .fsscript, .ml or .mli.
>
```

While the error message clearly states what the problem is here, I think there are a couple of issues to notice:
- It is common practice to omit file extensions in executables. Perhaps we should consider supporting this in the fsi context?
- An invalid file extension will cause fsi to run in interactive mode, which seems a bit bizarre. It actually took me a while to understand what was going on because of all the text that was being output.

**Comment by KevinRansom** [comment]
@eiriktsarpalis 
Yes I agree, restricting the extensions of files in this way is incorrect.

**Comment by eiriktsarpalis** [comment]
@KevinRansom Do you think we should lift this restriction for fsi only or the compiler in general?

**Comment by dsyme** [comment]
@eiriktsarpalis  Just for FSI, and probably only for one script.  I think the extension matters too much for the compiler, c.f. fs v. fsi v. fsx, shich all have different rules.

**Comment by nodakai** [comment]
I tried installing an F# script to replace `/usr/bin/sendmail` (wanted to forward cron failure mails to webhooks) only to discover this restriction. i ended up installing a Bash wrapper which calls the `.fsx` script but I felt that was an unnecessary indirection.

I wouldn’t mind passing an extra argument like `--extension fsx` to `dotnet fsi` if that helps

**Comment by T-Gro** [comment]
If we speak of `dotnet fsi` only, the extension restriction could be lifted.
I will mark it as help wanted - if anyone is interested into contribution, feel free to reach out and discuss possibilities.

**Comment by ThobiWanKenobi** [comment]
Is this still a bug that exists? This issue is ancient, and if this has not been fixed yet, I will gladly contribute and see if we can get this one closed. 

If anyone has some information that would be beneficial for this problem I would appreciate it. By some simple investigation I can see there is an fsi project, which is where I assume we I should be looking first?

**Comment by Martin521** [comment]
I guess you have to try it out. The comment above seems to say it still existed a few months ago.
And yes, I believe the fsi project is the right starting point.

**Comment by ThobiWanKenobi** [comment]
Alright so, I have done some initial investigation on this topic.

1. I can confirm that this issue is very much still a thing, however, to keep this updated to use dotnet fsi the first line would be slight adjusted to contain #!/usr/bin/env -S dotnet fsi
2. I tried looking up where the error message comes from, and it comes from a generic "ParseAndCheckInputs" module. The way it currently works is that parsing needs to be differentiated based on the file extension such as implementation files, signature files, and these "ml" file extensions. 

I was thinking that since we are invoking from the fsi, it should be trivially easy to say that our parsing should be based on the implementation files, as it wouldn't make sense in an fsi session to use a signature file or these "ml" files? I am still considering how to create a proper solution for this, but this is what I have found currently.

**Comment by Martin521** [comment]
I don't think that's correct.
You can call `dotnet fsi` with both script and non-script files.
`dotnet fsi test.fs` compiles `test.fs` (going through `ParseAndCheckInputs`, calling `Parser.implementationFile`), loads it and goes to interactive mode.
`dotnet fsi test.fsx` executes the script `test.fsx` and exits (no interactive mode). The script is compiled by `fsi` through a different path (not `ParseAndCheckInputs`), calling `Parser.interaction`. So, using the F# script grammar (i.e. no modules etc).
The decision on the file suffix is made [here](https://github.com/dotnet/fsharp/blob/616530c723b1017697f4b85a3f76432c3b6881e4/src/Compiler/Interactive/fsi.fs#L1180), based on [this](https://github.com/dotnet/fsharp/blob/616530c723b1017697f4b85a3f76432c3b6881e4/src/Compiler/Driver/CompilerConfig.fs#L48) list. Possibly, just adding an empty string to that list would suffice.

**Comment by ThobiWanKenobi** [comment]
Well, that is quite interesting! That might explain. If you look in the attached image, I was looking at the fsi project? Which only contains two files, is that even used anymore then? How I came to the conclusion from my previous comment was search through the fsharp codebase for where the error is used, and the only thing I could find at least comes from that ParseAndCheckInputs file. Thanks for the tip! I will continue to investigate into this and see what I can find.

<img width="414" height="533" alt="Image" src="https://github.com/user-attachments/assets/bcb4e612-566e-4d56-81c5-c08e42845f8a" />

**Comment by Martin521** [comment]
Yes, the `fsi` project is just about creating the fsi executable (like `fsc` for the compiler). Most of the functionality is in F.C.S

**Comment by T-Gro** [comment]
@ThobiWanKenobi : I would recommend to start by finding the source of the error `FS0226`, and reverse engineering from there.

**Comment by ThobiWanKenobi** [comment]
Thanks for the recommendation. Also, I think it should be fitting to ask, is there any way, to execute the fsi from this build? You know invoke the command on my linux machine to test whether the changes I make work or not? I know how to use the VisualFsharp solution, but that is the IDE, and I don't know how to invoke our took from the terminal? Does anyone know?

**Comment by T-Gro** [comment]
Locate the mentioned `fsi` project (which creates its outputs in `artifacts` folder)
And use it in that folder with `dotnet run` on the project.
Or you could navigate to the `artifacts` folder, locate the fsi project there, and `dotnet fsi.dll` on the built binary, that should also work.

**Comment by Martin521** [comment]
@ThobiWanKenobi 
If you are using VS Code, you can also select "Launch FSI" in the debug window and run/debug FSI with Ctrl F5 / F5.
